### PR TITLE
Send old inputs to diff and update

### PR DESCRIPTION
**Pull request by Frassle**
# Description

Fixes https://github.com/pulumi/pulumi/issues/5973

Both Azure and Kubernetes set an extra field on created resource outputs (called "__inputs") that doesn't exist in the inputs to the corresponding Pulumi resources. Their Diff logic needs the subset of inputs that are managed by Pulumi to compare these values to live state in subsequent updates, while not showing diffs for fields that we don't manage directly. This change adds that data to the RPC protocol directly so these providers no longer need to manage these extra "__input" fields.

This adds two new properties to the RPC interfaces with providers. Firstly we pass a new bool flag to Configure to tell the provider if it will get sent old inputs as part of Diff. Secondly we now pass the old inputs and the old outputs to Diff and Update as opposed to just the old outputs.
    
The flag passed to Configure isn't strictly needed. Providers should be able to tell that the old inputs aren't being sent to Diff and Updae by virtue of them being `null/nil/None` rather than an empty map. For DiffConfig this is the only way to detect this case because it's called before Configure. But its still useful to send this explictly to Configure to allow providers to return a configure error that they need to be ran against a newer engine version.

There is one fairly major failure case that can happen with this feature and providers who update to make use of this feature should ensure they call this out in their changelog.
That is if a provider has been saving old inputs in outputs as a way to get access to them in diff, and then updates to use these new protocol properties instead and stops saving the inputs in outputs _and then_ the user downgrades back to an old provider version things will probably get _very_ confused.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by Frassle** [comment]
cc @t0yv0 

I think this might also need to send old inputs to Create/Update for terraform bridge, but I wasn't sure?

File: proto/pulumi/provider.proto
Comment (t0yv0) [reply proto/pulumi/provider.proto:195]: Hm, perhaps it's just a vocab confusion between TF and Pulumi vocabulary. Perhaps the Pulumi vocabulary "old output values" essentially translates to "prior state" in the TF world, since TF bridge always publishes as Pulumi "outputs" all properties; the schema for output properties always includes every input property. Yeah this is probably fine.

Comment (t0yv0) [reply proto/pulumi/provider.proto:195]: Right, right. So in the Pulumi language it's properly called "outputs". It's good then.

**Comment by t0yv0** [comment]
This change LGTM from bridge POV it doesn't necessitate any major changes in the bridge but opens up opportunities to use the old_inputs for more precision down the line if needed.

**Comment by t0yv0** [comment]
Checking the code again, actually unlikely bridge will need to use this at all, so no problem there.

File: pkg/resource/deploy/providers/registry.go
Comment (lblackstone) [inline pkg/resource/deploy/providers/registry.go:322]: Is this result no longer needed?

**Comment by Frassle** [comment]
> The PR description is missing the why of the change. What problems will this solve?

Ah fair, I keep doing this on the basis that the linked issue lists the reason. @danielrbradley and @lblackstone might be able to give a better description than me, but I'll update the description with a summary of https://github.com/pulumi/pulumi/issues/5973.

**Review by lblackstone — APPROVED** [review]
🎉

**Comment by lblackstone** [comment]
> might be able to give a better description than me

For a little added context about why this is needed, both Azure and Kubernetes set a bunch of extra fields on the created resources that don't exist in the inputs to the corresponding Pulumi resources. The Diff logic needs the subset of inputs that are managed by Pulumi to compare these values to live state in subsequent updates.

**Comment by Frassle** [comment]
> @Frassle Oops, this might be on me for overlooking the linked issue. I think that's a fair approach.

No no, it's fair. You shouldn't have to read a whole issue thread to get the summary context of a change :)

File: pkg/resource/deploy/providers/registry.go
Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go:322]: It just falls through into doing this below + handling the default diffing we do for resources. We could add this back if we're worried that auto-diff on provider config might cause issues.

Comment (lblackstone) [reply pkg/resource/deploy/providers/registry.go:322]: Oh, I didn't see that this was identical to the error handling below. I think this should be fine.

**Comment by AaronFriel** [comment]
Definitely an improvement in this API - decouples "olds" from having to be a superset of news, which as we see with the providers mentioned isn't the case.

**Comment by lblackstone** [comment]
I just realized that this PR only updates `Diff` and `CheckDiff`. Should we make a similar change to `Update`?

**Comment by Frassle** [comment]
> I just realized that this PR only updates Diff and CheckDiff. Should we make a similar change to Update?

I wasn't sure, it would be quite easy to add in. Can't think of any complications it would cause down the line having to send old inputs to Update, if I haven't thought of any reason against it by Monday I'll slot it in.

**Comment by lblackstone** [comment]
LGTM!

**Review by danielrbradley — COMMENTED** [review]
I think we need a plan around how we roll this out given it's a breaking change to the Provider interface. 

I assume there isn't a way to avoid it being a breaking change given we can't do overloads in Go and it would be very messy to add new versions of every method to the interface?

If we merge right now what I expect will happen is that none of the automated provider update jobs will be able to pull the new version because the existing code will no longer compile. This will then require each provider to be updated manually. This might take some time and could mean that we're unable to pull in critical fixes until manually resolved.

On the change itself - this is great! We can also add the bridge to the list of providers that won't have to use `__inputs` any more.

**Comment by Frassle** [comment]
> I think we need a plan around how we roll this out given it's a breaking change to the Provider interface.

Yes :( We really really really ought to have migrated off this interface to a new one with extension points by now.

> I assume there isn't a way to avoid it being a breaking change given we can't do overloads in Go and it would be very messy to add new versions of every method to the interface?

https://github.com/pulumi/pulumi/issues/12010 has a design for a new interface built around struct arguments which would allow us to add things in a non-breaking way. We _could_ delay this until after we release that interface, but it's still going to require every provider to do a manual update to the new interface and will probably be a significant delay.

> We can also add the bridge to the list of providers that won't have to use __inputs any more.

I don't think terraform-bridge uses __inputs today, this change was for azure-native and k8s.

**Comment by Frassle** [comment]
> I think we need a plan around how we roll this out given it's a breaking change to the Provider interface.

cc @AaronFriel @kpitzen @justinvp about this, I know breaking changes were supposed to be on the cross-team discussion.

**Comment by t0yv0** [comment]
In case of bridged providers under ecosystem/ci-mgmt setup this can be mitigated because provider upgrade automation can upgrade the bridge version used, and this means that the sequence is fairly straightforward - one update in the bridge and then automatic updates work again. So there's a small bit of work but then it should roll, and as such this is a small friction change. Good point though that for all the other native + ci-mgmt providers the situation requires planned work.

**Comment by AaronFriel** [comment]
Can we fork the Provider interface and provide ProviderV2 for a non-breaking migration?

**Comment by Frassle** [comment]
> Can we fork the Provider interface and provide ProviderV2 for a non-breaking migration?

We could, but given this is the "sdk/go/common/resource/plugin".Provider interface I'm not really sure it's worth it. I think that's only used by terraform-bridge and can just be fixed up once.

Forking it either means making a totally separate ProviderV2 interface and also forking the Host type to have a HostV2 that returns that new interface. Or embedding Provider in ProviderV2 and having to implement Diff and DiffV2 and doing runtime casts from Host.Provider in the engine to get the ProviderV2 interface out.

The later is doable, but I'm not really sure it's worth it just to not break tfbridge.

**Comment by AaronFriel** [comment]
> We could, but given this is the "sdk/go/common/resource/plugin".Provider interface I'm not really sure it's worth it. I think that's only used by terraform-bridge and can just be fixed up once.

If that's the case, after this merges can we make a PR to TF to update that side of the interface?

**Comment by Frassle** [comment]
bors merge

**Comment by EronWright** [comment]
The concept of combining new inputs, old state, and old inputs is reminiscent of the [Kubernetes three-way merge patch calculation](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#how-apply-calculates-differences-and-merges-changes):
> The kubectl apply command calculates this patch request using the configuration file, the live configuration, and the last-applied-configuration annotation stored in the live configuration.

**Pull request merged**
